src/vanilla/utils/atomWithObservable.ts

```diff
@@ -81,107 +74,79 @@ export function atomWithObservable<Data>(
       observable = itself
     }
 
-    let resolve: ((result: Result) => void) | undefined
-    const makePending = () =>
-      new Promise<Result>((r) => {
-        resolve = r
-      })
-    const initialResult: Result | Promise<Result> =
-      options && 'initialValue' in options
-        ? {
-            d:
-              typeof options.initialValue === 'function'
-                ? (options.initialValue as () => Data)()
-                : (options.initialValue as Data),
-          }
-        : makePending()
-
-    let setResult: ((result: Result) => void) | undefined
-    let lastResult: Result | undefined
-    const listener = (result: Result) => {
-      lastResult = result
-      resolve?.(result)
-      setResult?.(result)
-    }
+    // TODO: replace with https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers
+    // once CI Node version is bump to 22+
+    let resolve: (value: Data | PromiseLike<Data>) => void
+    let reject: (reason?: any) => void
+    const promise = new Promise<Data>((_resolve, _reject) => {
+      resolve = _resolve
+      reject = _reject
+    })
 
-    let subscription: Subscription | undefined
-    let timer: Timeout | undefined
-    const isNotMounted = () => !setResult
-    const start = () => {
-      if (subscription) {
-        clearTimeout(timer)
-        subscription.unsubscribe()
-      }
-      subscription = observable.subscribe({
-        next: (d) => listener({ d }),
-        error: (e) => listener({ e }),
-        complete: () => {},
-      })
-      if (isNotMounted() && options?.unstable_timeout) {
-        timer = setTimeout(() => {
-          if (subscription) {
-            subscription.unsubscribe()
-            subscription = undefined
-          }
-        }, options.unstable_timeout)
-      }
-    }
-    start()
-
-    const resultAtom = atom(lastResult || initialResult)
+    const initialResult: Data | Promise<Data> =
+      options && 'initialValue' in options
+        ? typeof options.initialValue === 'function'
+          ? (options.initialValue as () => Data)()
+          : (options.initialValue as Data)
+        : promise
+
+    const resultAtom = atom(initialResult)
+
+    let preMountData: Data | typeof EMPTY = EMPTY
+
+    const initialSubscription = observable.subscribe({
```

Comment:
@dai-shi the extra subscription is here.

Reply:
Thread continue here 👉 https://github.com/pmndrs/jotai/pull/2631/files#r1650220462